File: services/orchestrator-core/internal/store/desktop_apps.go
package store

import (
	"context"
	"errors"
	"html"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

const maxDesktopApps = 1000

var plistStringPattern = regexp.MustCompile(`(?s)<key>([^<]+)</key>\s*<string>([^<]*)</string>`)

func executeDesktopAppList(ctx context.Context, request CapabilityRequest) (*CapabilityExecutionResult, error) {
	if request.Type == "" {
		request.Type = "capability_request"
	}
	if request.Risk == "" {
		request.Risk = "read_only"
	}
	if normalizedRisk(request.Risk) != "read_only" {
		return nil, ErrPolicyDenied
	}
	apps := listDesktopApps(ctx)
	return &CapabilityExecutionResult{
		CapabilityRequest: request,
		PolicyDecision:    map[string]any{"risk": "read_only", "decision": "allow", "reason": "desktop_app_list_v1 reads only local app bundle metadata"},
		Workflow: ToolWorkflow{
			WorkflowName: "desktop_app_list_v1",
			Capability:   "desktop_app_list",
			RiskLevel:    "read_only",
			Steps:        []ToolWorkflowStep{{Tool: "desktop_list_app_bundles", Args: map[string]any{}, RiskLevel: "read_only"}},
		},
		SelectedNodeID: "main-node",
		NormalizedResult: map[string]any{
			"status": "completed",
			"total":  len(apps),
			"apps":   apps,
			"mode":   "desktop_app_list_v1_bundle_scan",
		},
	}, nil
}

func executeDesktopAppInspect(ctx context.Context, request CapabilityRequest) (*CapabilityExecutionResult, error) {
	if request.Type == "" {
		request.Type = "capability_request"
	}
	if request.Risk == "" {
		request.Risk = "read_only"
	}
	if normalizedRisk(request.Risk) != "read_only" {
		return nil, ErrPolicyDenied
	}
	name := strings.ToLower(strings.TrimSpace(stringInput(request.Inputs, "name", "")))
	bundleID := strings.ToLower(strings.TrimSpace(stringInput(request.Inputs, "bundle_id", "")))
	targetPath := strings.TrimSpace(stringInput(request.Inputs, "path", ""))
	if name == "" && bundleID == "" && targetPath == "" {
		return nil, ErrMissingArgument
	}
	apps := listDesktopApps(ctx)
	matches := []map[string]any{}
	for _, app := range apps {
		appName := strings.ToLower(stringInput(app, "name", ""))
		appBundleID := strings.ToLower(stringInput(app, "bundle_id", ""))
		appPath := stringInput(app, "path", "")
		if (name != "" && strings.Contains(appName, name)) ||
			(bundleID != "" && appBundleID == bundleID) ||
			(targetPath != "" && filepath.Clean(appPath) == filepath.Clean(targetPath)) {
			matches = append(matches, app)
		}
	}
	return &CapabilityExecutionResult{
		CapabilityRequest: request,
		PolicyDecision:    map[string]any{"risk": "read_only", "decision": "allow", "reason": "desktop_app_inspect_v1 reads only one app bundle metadata"},
		Workflow: ToolWorkflow{
			WorkflowName: "desktop_app_inspect_v1",
			Capability:   "desktop_app_inspect",
			RiskLevel:    "read_only",
			Steps:        []ToolWorkflowStep{{Tool: "desktop_inspect_app_bundle", Args: map[string]any{}, RiskLevel: "read_only"}},
		},
		SelectedNodeID: "main-node",
		NormalizedResult: map[string]any{
			"status":  "completed",
			"matches": matches,
			"total":   len(matches),
			"mode":    "desktop_app_inspect_v1_bundle_scan",
		},
	}, nil
}

func listDesktopApps(ctx context.Context) []map[string]any {
	roots := desktopAppRoots()
	seen := map[string]bool{}
	apps := []map[string]any{}
	for _, root := range roots {
		if len(apps) >= maxDesktopApps || ctx.Err() != nil {
			break
		}
		info, err := os.Stat(root.path)
		if err != nil || !info.IsDir() {
			continue
		}
		_ = filepath.WalkDir(root.path, func(path string, entry os.DirEntry, walkErr error) error {
			if walkErr != nil || ctx.Err() != nil || len(apps) >= maxDesktopApps {
				return nil
			}
			if path != root.path && entry.IsDir() && strings.HasSuffix(entry.Name(), ".app") {
				clean := filepath.Clean(path)
				if !seen[clean] {
					seen[clean] = true
					apps = append(apps, appBundleMetadata(clean, root.source))
				}
				return filepath.SkipDir
			}
			if entry.IsDir() && appSearchDepth(root.path, path) > 3 {
				return filepath.SkipDir
			}
			return nil
		})
	}
	sort.SliceStable(apps, func(i, j int) bool {
		return strings.ToLower(stringInput(apps[i], "name", "")) < strings.ToLower(stringInput(apps[j], "name", ""))
	})
	return apps
}

type desktopAppRoot struct {
	path   string
	source string
}

func desktopAppRoots() []desktopAppRoot {
	roots := []desktopAppRoot{
		{path: "/Applications", source: "applications"},
		{path: "/System/Applications", source: "system"},
	}
	if home, err := os.UserHomeDir(); err == nil && strings.TrimSpace(home) != "" {
		roots = append(roots, desktopAppRoot{path: filepath.Join(home, "Applications"), source: "user"})
	}
	return roots
}

func appSearchDepth(root string, path string) int {
	rel, err := filepath.Rel(root, path)
	if err != nil || rel == "." {
		return 0
	}
	return len(strings.Split(rel, string(filepath.Separator)))
}

func appBundleMetadata(path string, source string) map[string]any {
	infoPath := filepath.Join(path, "Contents", "Info.plist")
	plist := readPlistStrings(infoPath)
	name := strings.TrimSuffix(filepath.Base(path), ".app")
	if displayName := firstNonEmptyString(plist["CFBundleDisplayName"], plist["CFBundleName"]); displayName != "" {
		name = displayName
	}
	return map[string]any{
		"name":             name,
		"path":             path,
		"source":           source,
		"bundle_id":        plist["CFBundleIdentifier"],
		"version":          firstNonEmptyString(plist["CFBundleShortVersionString"], plist["CFBundleVersion"]),
		"executable":       plist["CFBundleExecutable"],
		"metadata_source":  plist["metadata_source"],
		"content_readable": false,
	}
}

func readPlistStrings(path string) map[string]string {
	result := map[string]string{"metadata_source": "bundle_path"}
	raw, err := os.ReadFile(path)
	if err != nil || len(raw) == 0 {
		return result
	}
	if len(raw) > 512*1024 {
		raw = raw[:512*1024]
	}
	if strings.HasPrefix(string(raw[:min(len(raw), 6)]), "bplist") {
		result["metadata_source"] = "binary_plist_unparsed"
		return result
	}
	for _, match := range plistStringPattern.FindAllStringSubmatch(string(raw), -1) {
		if len(match) == 3 {
			result[html.UnescapeString(strings.TrimSpace(match[1]))] = html.UnescapeString(strings.TrimSpace(match[2]))
		}
	}
	result["metadata_source"] = "info_plist"
	return result
}

func firstNonEmptyString(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return strings.TrimSpace(value)
		}
	}
	return ""
}

func min(a int, b int) int {
	if a < b {
		return a
	}
	return b
}

func errNoDesktopAppMatch() error {
	return errors.New("desktop app not found")
}

File: packages/runtime/src/desktop-apps.ts
import { readdirSync, readFileSync, statSync } from 'node:fs';
import { homedir } from 'node:os';
import { basename, join, normalize, relative } from 'node:path';
import type { CapabilityResult } from './capabilities.ts';

export type DesktopAppListRequest = {
  max_results?: unknown;
};

export type DesktopAppInspectRequest = {
  name?: string;
  bundle_id?: string;
  path?: string;
};

export type DesktopAppRoot = {
  path: string;
  source: string;
};

type DesktopAppMetadata = {
  name: string;
  path: string;
  source: string;
  bundle_id: string;
  version: string;
  executable: string;
  metadata_source: string;
  content_readable: false;
};

const defaultMaxDesktopApps = 1000;
const plistStringPattern = /<key>([^<]+)<\/key>\s*<string>([^<]*)<\/string>/gs;

export function executeDesktopAppList(
  req: DesktopAppListRequest = {},
  roots: DesktopAppRoot[] = desktopAppRoots(),
): CapabilityResult {
  const apps = listDesktopApps(roots, boundedAppLimit(req.max_results));
  return {
    status: 'completed',
    total: apps.length,
    apps,
    summary: `Found ${apps.length} local app bundle(s).`,
    mode: 'desktop_app_list_v1_bundle_scan',
  };
}

export function executeDesktopAppInspect(
  req: DesktopAppInspectRequest,
  roots: DesktopAppRoot[] = desktopAppRoots(),
): CapabilityResult {
  const name = req.name?.trim().toLowerCase() || '';
  const bundleID = req.bundle_id?.trim().toLowerCase() || '';
  const targetPath = req.path?.trim() || '';
  if (!name && !bundleID && !targetPath) throw new Error('desktop_app_inspect requires name, bundle_id, or path');
  const matches = listDesktopApps(roots, defaultMaxDesktopApps).filter((app) => {
    const appName = app.name.toLowerCase();
    const appBundleID = app.bundle_id.toLowerCase();
    return (name && appName.includes(name))
      || (bundleID && appBundleID === bundleID)
      || (targetPath && normalize(app.path) === normalize(targetPath));
  });
  return {
    status: 'completed',
    total: matches.length,
    matches,
    summary: matches.length > 0
      ? `Found ${matches.length} matching app bundle(s).`
      : 'No matching local app bundle found.',
    mode: 'desktop_app_inspect_v1_bundle_scan',
  };
}

function listDesktopApps(roots: DesktopAppRoot[], limit: number): DesktopAppMetadata[] {
  const seen = new Set<string>();
  const apps: DesktopAppMetadata[] = [];
  for (const root of roots) {
    if (apps.length >= limit) break;
    try {
      if (!statSync(root.path).isDirectory()) continue;
    } catch {
      continue;
    }
    walkAppRoot(root.path, root.path, root.source, apps, seen, limit);
  }
  apps.sort((a, b) => a.name.toLowerCase().localeCompare(b.name.toLowerCase()));
  return apps;
}

function walkAppRoot(root: string, current: string, source: string, apps: DesktopAppMetadata[], seen: Set<string>, limit: number): void {
  if (apps.length >= limit) return;
  let entries;
  try {
    entries = readdirSync(current, { withFileTypes: true });
  } catch {
    return;
  }
  for (const entry of entries) {
    if (!entry.isDirectory() || apps.length >= limit) continue;
    const path = join(current, entry.name);
    if (entry.name.endsWith('.app')) {
      const normalized = normalize(path);
      if (!seen.has(normalized)) {
        seen.add(normalized);
        apps.push(appBundleMetadata(normalized, source));
      }
      continue;
    }
    if (appSearchDepth(root, path) <= 3) walkAppRoot(root, path, source, apps, seen, limit);
  }
}

function desktopAppRoots(): DesktopAppRoot[] {
  return [
    { path: '/Applications', source: 'applications' },
    { path: '/System/Applications', source: 'system' },
    { path: join(homedir(), 'Applications'), source: 'user' },
  ];
}

function appBundleMetadata(path: string, source: string): DesktopAppMetadata {
  const plist = readPlistStrings(join(path, 'Contents', 'Info.plist'));
  const name = firstNonEmptyString(plist.CFBundleDisplayName, plist.CFBundleName) || basename(path, '.app');
  return {
    name,
    path,
    source,
    bundle_id: plist.CFBundleIdentifier || '',
    version: firstNonEmptyString(plist.CFBundleShortVersionString, plist.CFBundleVersion),
    executable: plist.CFBundleExecutable || '',
    metadata_source: plist.metadata_source || 'bundle_path',
    content_readable: false,
  };
}

function readPlistStrings(path: string): Record<string, string> {
  const result: Record<string, string> = { metadata_source: 'bundle_path' };
  let raw: Buffer;
  try {
    raw = readFileSync(path);
  } catch {
    return result;
  }
  if (raw.length === 0) return result;
  if (raw.subarray(0, 6).toString('utf8') === 'bplist') {
    return { metadata_source: 'binary_plist_unparsed' };
  }
  const text = raw.subarray(0, 512 * 1024).toString('utf8');
  for (const match of text.matchAll(plistStringPattern)) {
    result[htmlUnescape(match[1].trim())] = htmlUnescape(match[2].trim());
  }
  result.metadata_source = 'info_plist';
  return result;
}

function appSearchDepth(root: string, path: string): number {
  const rel = relative(root, path);
  if (!rel || rel === '.') return 0;
  return rel.split(/[\\/]+/).filter(Boolean).length;
}

function boundedAppLimit(value: unknown): number {
  const number = Number(value);
  if (!Number.isFinite(number) || number <= 0) return defaultMaxDesktopApps;
  return Math.min(Math.floor(number), defaultMaxDesktopApps);
}

function firstNonEmptyString(...values: Array<string | undefined>): string {
  return values.map((value) => value?.trim() || '').find(Boolean) || '';
}

function htmlUnescape(value: string): string {
  return value
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'");
}
